File: mobile/shared/src/commonTest/kotlin/com/bissbilanz/analytics/CaloricLagTest.kt
package com.bissbilanz.analytics

import kotlinx.datetime.plus
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

private fun pad2(n: Int): String = n.toString().padStart(2, '0')

class CaloricLagTest {
    private fun makeSeries(
        dates: List<String>,
        values: List<Double?>,
    ): List<Pair<String, Double?>> = dates.zip(values)

    @Test
    fun insufficientPairsReturnsNullCorrelation() {
        val calDates = (1..5).map { "2024-01-${pad2(it)}" }
        val wtDates = (1..5).map { "2024-01-${pad2(it)}" }
        val calories = makeSeries(calDates, (1..5).map { 2000.0 })
        val weights = makeSeries(wtDates, (1..5).map { 80.0 })
        val result = computeCaloricLag(calories, weights, maxLag = 3)
        for (r in result.results) {
            assertNull(r.correlation)
        }
        assertNull(result.bestLag)
    }

    @Test
    fun clearLagPatternSelectsBestLag() {
        // Weight moves day over day in proportion to intake three days earlier;
        // intake is a fixed pseudo-random sequence so neighbouring lags do not
        // correlate with each other.
        val n = 30
        val lag = 3
        var seed = 42L

        fun rand(): Double {
            seed = (seed * 1664525L + 1013904223L) and 0xFFFFFFFFL
            return seed.toDouble() / 4294967296.0
        }
        val calories = (0 until n + lag).map { Pair(isoDay(it), (2000.0 + (rand() - 0.5) * 800) as Double?) }
        var w = 80.0
        val weights =
            (lag until n + lag).map { i ->
                w += (calories[i - lag].second!! - 2000.0) / 20000.0
                Pair(isoDay(i), w as Double?)
            }
        val result = computeCaloricLag(calories, weights, maxLag = 5)
        assertEquals(lag, result.bestLag)
        assertEquals(5, result.comparisons)
        val best = result.results.first { it.lag == lag }
        assertTrue(best.correlation!!.r > 0.99)
        assertTrue(best.qValue!! < 0.05)
    }

    @Test
    fun unrelatedRandomWalkYieldsNoBestLag() {
        var seed = 7L

        fun rand(): Double {
            seed = (seed * 1664525L + 1013904223L) and 0xFFFFFFFFL
            return seed.toDouble() / 4294967296.0
        }
        val calories = (0 until 40).map { Pair(isoDay(it), (2000.0 + (rand() - 0.5) * 800) as Double?) }
        var w = 80.0
        val weights =
            calories.map { (date, _) ->
                w += (rand() - 0.5) * 0.4
                Pair(date, w as Double?)
            }
        val result = computeCaloricLag(calories, weights, maxLag = 7)
        assertNull(result.bestLag)
    }

    private fun isoDay(i: Int): String =
        kotlinx.datetime
            .LocalDate(2024, 1, 1)
            .plus(i, kotlinx.datetime.DateTimeUnit.DAY)
            .toString()

    @Test
    fun allResultsHaveLagValues() {
        val n = 20
        val dates = (1..n).map { "2024-01-${pad2(it)}" }
        val calories = makeSeries(dates, (1..n).map { 2000.0 })
        val weights = makeSeries(dates, (1..n).map { 80.0 - it * 0.05 })
        val maxLag = 5
        val result = computeCaloricLag(calories, weights, maxLag = maxLag)
        assertEquals(maxLag, result.results.size)
        for (i in 0 until maxLag) {
            assertEquals(i + 1, result.results[i].lag)
        }
    }

    @Test
    fun bestLagHasHighestAbsCorrelation() {
        val n = 20
        val dates = (1..n).map { "2024-01-${pad2(it)}" }
        val calories = makeSeries(dates, (1..n).map { it.toDouble() * 100 + 1500 })
        val weights = makeSeries(dates, (1..n).map { 80.0 - it * 0.1 })
        val result = computeCaloricLag(calories, weights, maxLag = 3)
        val best = result.bestLag
        if (best != null) {
            val bestCorr = result.results.first { it.lag == best }.correlation
            assertNotNull(bestCorr)
            for (r in result.results) {
                if (r.correlation != null) {
                    assertTrue(abs(r.correlation.r) <= abs(bestCorr.r) + 1e-9)
                }
            }
        }
    }

    @Test
    fun emptyCaloriesReturnsAllNullCorrelations() {
        val weight = (1..10).map { Pair("2024-01-${pad2(it)}", 80.0 as Double?) }
        val result = computeCaloricLag(emptyList(), weight)
        assertEquals(7, result.results.size)
        assertNull(result.bestLag)
        for (r in result.results) assertNull(r.correlation)
    }

    @Test
    fun emptyWeightReturnsAllNullCorrelations() {
        val calories = (1..10).map { Pair("2024-01-${pad2(it)}", 2000.0 as Double?) }
        val result = computeCaloricLag(calories, emptyList())
        assertEquals(7, result.results.size)
        assertNull(result.bestLag)
    }

    @Test
    fun allNullCaloriesReturnsAllNullCorrelations() {
        val calories = (1..10).map { Pair("2024-01-${pad2(it)}", null as Double?) }
        val weight = (1..10).map { Pair("2024-01-${pad2(it)}", 80.0 as Double?) }
        val result = computeCaloricLag(calories, weight)
        assertNull(result.bestLag)
    }

    @Test
    fun maxLagOneReturnsSingleResult() {
        val calories = (1..10).map { Pair("2024-01-${pad2(it)}", 2000.0 as Double?) }
        val weight = (1..10).map { Pair("2024-01-${pad2(it)}", 80.0 as Double?) }
        val result = computeCaloricLag(calories, weight, maxLag = 1)
        assertEquals(1, result.results.size)
        assertEquals(1, result.results[0].lag)
    }
}

File: src/lib/analytics/__tests__/caloric-lag.test.ts
import { describe, it, expect } from 'vitest';
import { computeCaloricLag } from '../caloric-lag';

/** Deterministic pseudo-random intake so neighbouring lags are not autocorrelated. */
function lcg(seed: number) {
	let s = seed >>> 0;
	return () => {
		s = (Math.imul(1664525, s) + 1013904223) >>> 0;
		return s / 2 ** 32;
	};
}

function isoDay(startDate: string, i: number): string {
	const d = new Date(startDate + 'T00:00:00Z');
	d.setUTCDate(d.getUTCDate() + i);
	return d.toISOString().slice(0, 10);
}

/**
 * Weight moves day over day in proportion to intake `lag` days earlier, so the
 * signal lives in the weight *difference*, which is what the analytic tests.
 */
function makeCalorieWeight(
	startDate: string,
	n: number,
	lag: number,
	baseCalories: number = 2000,
	baseWeight: number = 80
) {
	const rand = lcg(42);
	const calories: { date: string; value: number }[] = [];
	const weight: { date: string; value: number }[] = [];

	for (let i = 0; i < n + lag; i++) {
		calories.push({ date: isoDay(startDate, i), value: baseCalories + (rand() - 0.5) * 800 });
	}

	let w = baseWeight;
	for (let i = lag; i < n + lag; i++) {
		w += (calories[i - lag].value - baseCalories) / 20000;
		weight.push({ date: isoDay(startDate, i), value: w });
	}

	return { calories, weight };
}

describe('computeCaloricLag', () => {
	it('identifies lag=3 as best when weight change follows intake by 3 days', () => {
		const { calories, weight } = makeCalorieWeight('2024-01-01', 30, 3);
		const result = computeCaloricLag(calories, weight, 5);
		expect(result.bestLag).toBe(3);
		const best = result.results.find((r) => r.lag === 3)!;
		expect(best.correlation!.r).toBeGreaterThan(0.99);
		expect(best.qValue).not.toBeNull();
		expect(best.qValue!).toBeLessThan(0.05);
	});

	it('reports how many lags were testable', () => {
		const { calories, weight } = makeCalorieWeight('2024-01-01', 30, 2);
		const result = computeCaloricLag(calories, weight, 7);
		expect(result.comparisons).toBe(7);
	});

	it('returns no best lag when nothing survives FDR control', () => {
		// Weight is an unrelated random walk.
		const rand = lcg(7);
		const calories = Array.from({ length: 40 }, (_, i) => ({
			date: isoDay('2024-01-01', i),
			value: 2000 + (rand() - 0.5) * 800
		}));
		let w = 80;
		const weight = calories.map((c) => {
			w += (rand() - 0.5) * 0.4;
			return { date: c.date, value: w };
		});
		const result = computeCaloricLag(calories, weight, 7);
		expect(result.bestLag).toBeNull();
		for (const r of result.results) {
			expect(r.qValue === null || r.qValue > 0.05).toBe(true);
		}
	});

	it('returns results array with one entry per lag offset', () => {
		const { calories, weight } = makeCalorieWeight('2024-01-01', 20, 2);
		const result = computeCaloricLag(calories, weight, 7);
		expect(result.results).toHaveLength(7);
		expect(result.results[0].lag).toBe(1);
		expect(result.results[6].lag).toBe(7);
	});

	it('needs consecutive weights to form a day-over-day change', () => {
		// Weights every other day never form a consecutive pair → nothing testable.
		const calories = Array.from({ length: 16 }, (_, i) => ({
			date: isoDay('2024-01-01', i),
			value: 2000 + i
		}));
		const weight = Array.from({ length: 8 }, (_, i) => ({
			date: isoDay('2024-01-01', i * 2),
			value: 80 + i * 0.1
		}));
		const result = computeCaloricLag(calories, weight, 3);
		expect(result.comparisons).toBe(0);
		expect(result.bestLag).toBeNull();
		for (const r of result.results) expect(r.correlation).toBeNull();
	});

	it('returns null correlation when too few paired data points for a lag', () => {
		const calories = [
			{ date: '2024-01-01', value: 2000 },
			{ date: '2024-01-02', value: 2100 },
			{ date: '2024-01-03', value: 2200 }
		];
		const weight = [
			{ date: '2024-01-04', value: 80.0 },
			{ date: '2024-01-05', value: 80.1 }
		];
		const result = computeCaloricLag(calories, weight, 3);
		for (const r of result.results) {
			expect(r.correlation).toBeNull();
		}
	});

	it('returns bestLag=null when all offsets have too few data points', () => {
		const calories = [
			{ date: '2024-01-01', value: 2000 },
			{ date: '2024-01-02', value: 2100 }
		];
		const weight = [
			{ date: '2024-01-05', value: 80.0 },
			{ date: '2024-01-06', value: 80.1 }
		];
		const result = computeCaloricLag(calories, weight, 3);
		expect(result.bestLag).toBeNull();
	});

	it('handles null values in input series', () => {
		const { calories, weight } = makeCalorieWeight('2024-01-01', 20, 2);
		const withNulls = calories.map((c, i) => (i % 5 === 0 ? { ...c, value: null } : c));
		const result = computeCaloricLag(withNulls, weight, 2);
		expect(result.results).toHaveLength(2);
	});

	it('uses default maxLag of 7', () => {
		const { calories, weight } = makeCalorieWeight('2024-01-01', 25, 2);
		const result = computeCaloricLag(calories, weight);
		expect(result.results).toHaveLength(7);
	});

	it('returns empty results for maxLag=0', () => {
		const calories = [
			{ date: '2024-01-01', value: 2000 },
			{ date: '2024-01-02', value: 2500 }
		];
		const weight = [
			{ date: '2024-01-01', value: 80 },
			{ date: '2024-01-02', value: 80.1 }
		];
		const result = computeCaloricLag(calories, weight, 0);
		expect(result.bestLag).toBeNull();
		expect(result.comparisons).toBe(0);
		expect(result.results).toHaveLength(0);
	});

	it('handles constant weight series — every change is zero, constant input', () => {
		const { calories } = makeCalorieWeight('2024-01-01', 20, 2);
		const weight = calories.slice(2).map((c) => ({ date: c.date, value: 80.0 }));
		const result = computeCaloricLag(calories, weight, 3);
		expect(result.bestLag).toBeNull();
		for (const r of result.results) {
			if (r.correlation !== null) {
				expect(r.correlation.constantInput).toBe(true);
				expect(r.correlation.r).toBe(0);
			}
		}
	});
});
